pyemma/coordinates/io/featurizer.py

```diff
@@ -67,45 +67,158 @@ def __init__(self, func=None, *args, **kwargs):
     def describe(self):
         return "override me to get proper description!"
 
+    @property
+    def dimension(self):
+        # override me to get correct dimension
+        return 0
+
     def map(self, traj):
         feature = self._func(traj, self._args, self._kwargs)
         assert isinstance(feature, np.ndarray)
         return feature
 
 
+class SelectionFeature:
+    """
+    Just provide the cartesian coordinates of a selection of atoms (could be simply all atoms).
+    The coordinates are flattened as follows: [x1, y1, z1, x2, y2, z2, ...]
+
+    """
+    #TODO: Needs an orientation option
+
+    def __init__(self, top, indexes):
+        self.top = top
+        self.indexes = np.array(indexes)
+        self.prefix_label = "ATOM:"
+
+    def describe(self):
+        labels = []
+        for i in self.indexes:
+            labels.append("%s%s" % (self.prefix_label, _describe_atom(self.top, i)))
+        return labels
+
+    @property
+    def dimension(self):
+        return 3*self.indexes.shape[0]
+
+    def map(self, traj):
+        newshape = (traj.xyz.shape[0], 3*self.indexes.shape[0])
+        return np.reshape(traj.xyz[:,self.indexes,:], newshape)
+
+
 class DistanceFeature:
 
-    def __init__(self, top, distance_indexes):
+    def __init__(self, top, distance_indexes, periodic = True):
         self.top = top
         self.distance_indexes = np.array(distance_indexes)
         self.prefix_label = "DIST:"
+        self.periodic = periodic
 
     def describe(self):
         labels = []
         for pair in self.distance_indexes:
-            labels.append("%s%s - %s" % (self.prefix_label,
-                                         _describe_atom(self.top, pair[0]),
-                                         _describe_atom(self.top, pair[1])))
+            labels.append("%s %s - %s" % (self.prefix_label,
+                                          _describe_atom(self.top, pair[0]),
+                                          _describe_atom(self.top, pair[1])))
         return labels
 
+    @property
+    def dimension(self):
+        return self.distance_indexes.shape[0]
+
     def map(self, traj):
-        return mdtraj.compute_distances(traj, self.distance_indexes)
+        return mdtraj.compute_distances(traj, self.distance_indexes, periodic=self.periodic)
 
 
 class InverseDistanceFeature(DistanceFeature):
 
-    def __init__(self, top, distance_indexes):
-        DistanceFeature.__init__(self, top, distance_indexes)
+    def __init__(self, top, distance_indexes, periodic = True):
+        DistanceFeature.__init__(self, top, distance_indexes, periodic=periodic)
         self.prefix_label = "INVDIST:"
 
     def map(self, traj):
-        return 1.0 / mdtraj.compute_distances(traj, self.distance_indexes)
+        return 1.0 / mdtraj.compute_distances(traj, self.distance_indexes, periodic=self.periodic)
+
+
+class ContactFeature(DistanceFeature):
```

Comment:
Am 29/03/15 um 16:01 schrieb Martin K. Scherer:
>
> In pyemma/coordinates/io/featurizer.py 
> <https://github.com/markovmodel/PyEMMA/pull/162#discussion_r27354664>:
>
> > +class ContactFeature(DistanceFeature):
>
> why not use 
> http://mdtraj.org/latest/api/generated/mdtraj.compute_contacts.html#mdtraj.compute_contacts?
>
mdtraj.compute_contact seems to compute distances too rather than 
distances. At least the implementation I have found in the devel branch 
did not compute contacts. I have nothing against changing it if the 
behavior stays the same.


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------